### DISCUSSION TITLE
Register custom server?

### DISCUSSION THREAD

**Discussion by alex-popov-tech** [mason-org/mason.nvim#189] [Q&A] [answered]
Hello @williamboman , as always thank you for such a beautiful software you are building <3

Question: I had an `ls_emmet` server setup like that:
```
local custom_servers = {
    {
        server_name = 'ls_emmet',
        async = true,
        lspconfig = {
            default_config = {
                filetypes = {
                    'html',
                    'typescriptreact',
                    'javascriptreact',
                    'css'
                },
                root_dir = function()
                    return lspconfig.util.find_git_ancestor(vim.fn.getcwd()) or
                               lspconfig.util
                                   .find_package_json_ancestor(vim.fn.getcwd())
                end
            }
        },
        installer_server = server.Server:new{
            name = 'ls_emmet',
            async = true,
            root_dir = server.get_server_root_path('ls_emmet'),
            installer = npm.packages {'ls_emmet'},
            default_options = {
                cmd_env = npm.env(server.get_server_root_path('ls_emmet')),
                cmd = {
                    server.get_server_root_path('ls_emmet') ..
                        '/node_modules/.bin/ls_emmet',
                    '--stdio'
                }
            }
        }
    }
}

for _, config in pairs(custom_servers) do
    configs[config.server_name] = config.lspconfig
    servers.register(config.installer_server)
end
```

and it was fine, but now im not sure how to do same with `mason`, can you please assist where to look?

**Comment by williamboman** [comment, marked as answer]
Hey man, thanks 😊!  Right now there's no provided API to add a custom package, I want to let things settle first, and I'd like to think it through before adding it. You can hack thinks together by modifying the `mason-registry.index` module directly, like

```lua
local index = require("mason-registry.index")
index["ls_emmet"] = "my.ls_emmet.module"
```

The `"my.ls_emmet.module"` needs to be a Lua module available in your runtime path (meaning `require("my.ls_emmet.module")` works) that returns an instance of a Mason package (`mason-core.package`). This is not guaranteed to work in the future, but it's the only way to achieve it today.

As for `ls_emmet`, is there any reason you use it over `emmet-ls`? I wouldn't be opposed to adding `ls_emmet` to Mason if it provides additional value over `emmet-ls`

**Reply by alex-popov-tech** [reply]
@williamboman i think we discussed that once here https://github.com/williamboman/nvim-lsp-installer/discussions/433

thanks for explanation, i will wait until official support then :D

**Comment by RaafatTurki** [comment]
Hey, any progress on this?
I'm down to help implementing the API if a schema is already drawn for it

My use case relates more into mason-lspconfig but the discussion is here.
Anyways, there are a few LSP servers that come bundled as a smaller part within a larger piece of software such as:
- [`dartls`](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#dartls) within `dart`
- [`gdscript`](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#gdscript) within `godot`

My ideal end goal was being able to `:LspInstall` in a dart file and getting prompted my custom `dartls` package which checks for the existence of `dart` and if so it would be considered installed

My monkey patching attempts are far less ambitious from those ideal goals, I would settle for just being able to call 
```lua
require 'mason-lspconfig'.setup_handlers {
  dartls = function()
    require 'lspconfig'.dartls.setup {}
  end
}
```

The answer above only takes me half way in
```lua
-- adding local packages to mason
require('mason-registry.index')['dartls'] = 'extras.mason_local_reg.dartls'

-- adding local packages to mason-lspconfig
require 'mason-lspconfig.mappings.server'.lspconfig_to_package['dartls'] = 'dartls'
```
where `extras.mason_local_reg.dartls` is package that look like
```lua
return Pkg.new {
    name = "dartls",
    desc = [[Dummy package for dartls]],
    homepage = "https://some/link.com",
    languages = { Pkg.Lang.Dart },
    categories = { Pkg.Cat.LSP },
    install = function(ctx)
        ctx.receipt:with_primary_source({})
    end,
}
```
and it's missing many niceties such as filetype detection, command autocompletion.
and above it all, it doesn't execute the `dartls` handler so this was all for nothing :(
